File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_EMA_LastBarCrosses.java
package com.kishlaly.ta.analyze.tasks.blocks.one;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.EMAJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.utils.CollectionUtilsJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.LAST_QUOTE_NOT_CROSSING_EMA_SCREEN_1;
import static com.kishlaly.ta.model.BlockResultCodeJava.OK;
import static com.kishlaly.ta.utils.Quotes.isQuoteCrossedEMA;

/**
 * the last bar crosses the EMA26
 */
public class Long_ScreenOne_EMA_LastBarCrosses implements ScreenOneBlock {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<EMAJava> screen_1_EMA26 = (List<EMAJava>) screen.indicators.get(IndicatorJava.EMA26);
        if (!isQuoteCrossedEMA(screen.getLastQuote(), CollectionUtilsJava.getFromEnd(screen_1_EMA26, 1).getValue())) {
            LogJava.recordCode(LAST_QUOTE_NOT_CROSSING_EMA_SCREEN_1, screen);
            LogJava.addDebugLine("The last bar does not cross the EMA on the long term screen");
            return new BlockResultJava(screen.getLastQuote(), LAST_QUOTE_NOT_CROSSING_EMA_SCREEN_1);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/one/Long_ScreenOne_EMA_LastBarCrosses.kt
package com.kishlaly.ta.analyze.tasks.blocks.one

import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.EMA
import com.kishlaly.ta.model.indicators.Indicator
import com.kishlaly.ta.model.isCrossesEMA
import com.kishlaly.ta.utils.Log

/**
 * the last bar crosses the EMA26
 */
class Long_ScreenOne_EMA_LastBarCrosses : ScreenOneBlock {
    override fun check(screen: SymbolData): BlockResult {
        val screen_1_EMA26 = screen.indicators[Indicator.EMA26] as List<EMA>
        if (!screen.lastQuote.isCrossesEMA(screen_1_EMA26.last().value)) {
            Log.recordCode(BlockResultCode.LAST_QUOTE_NOT_CROSSING_EMA_SCREEN_1, screen)
            Log.addDebugLine("The last bar does not cross the EMA on the long term screen")
            return BlockResult(screen.lastQuote, BlockResultCode.LAST_QUOTE_NOT_CROSSING_EMA_SCREEN_1)
        }

        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}